File: app/src/main/java/jp/juggler/fadownloader/WorkerBase.java
package jp.juggler.fadownloader;

import java.util.concurrent.atomic.AtomicReference;

abstract public class WorkerBase extends Thread implements CancelChecker{

	public synchronized void waitEx( long ms ){
		try{
			wait( ms );
		}catch( InterruptedException ignored ){
		}
	}

	public synchronized void notifyEx(){
		notify();
	}

	final AtomicReference<String> cancel_reason = new AtomicReference<>( null );

	@Override public boolean isCancelled(){
		return cancel_reason.get() != null;
	}

	public boolean cancel(String reason){
		boolean rv = cancel_reason.compareAndSet( null, reason );
		notifyEx();
		return rv;
	}

	public abstract void run();
}

File: app/src/main/java/jp/juggler/fadownloader/util/WorkerBase.kt
package jp.juggler.fadownloader.util

import jp.juggler.fadownloader.CancelChecker
import java.util.concurrent.atomic.AtomicReference

abstract class WorkerBase : Thread(), CancelChecker {
	
	internal val cancel_reason = AtomicReference<String>(null)
	
	override val isCancelled : Boolean
		get() = cancel_reason.get() != null
	
	@Synchronized
	fun waitEx(ms : Long) {
		try {
			WaitHelper.wait(this,ms)
		} catch(ignored : InterruptedException) {
		}
		
	}
	
	@Synchronized
	fun notifyEx() {
		WaitHelper.notify(this)
	}
	
	open fun cancel(reason : String) : Boolean {
		val rv = cancel_reason.compareAndSet(null, reason)
		notifyEx()
		return rv
	}
	
	abstract override fun run()
}
